src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs

```diff
@@ -54,7 +54,13 @@ or CollectionType.IEnumerable
             // a single Array.Clone / cast mapping call should be sufficient and fast,
             // use a for loop mapping otherwise.
             if (!elementMapping.IsSynthetic)
-                return new ArrayForMapping(ctx.Source, ctx.Target, elementMapping, elementMapping.TargetType);
+            {
+                // upgrade nullability of element type
+                var targetValue = ctx.Target.IsArrayType()
```

Comment:
Couldn't this be simplified to `ctx.CollectionInfos.Target.CollectionType == CollectionType.Array`?